platform/linux-generic/arch/aarch64/odp_atomic.h

```diff
@@ -218,4 +217,103 @@ static inline __int128 __lockfree_load_16(__int128 *var, int mo)
+ * Helper macro for lockless atomic CAS operations on 128-bit integers
+ * @param[in,out] atom Pointer to the 128-bit atomic variable
+ * @param oper CAS operation
+ * @param old_val Old value
+ * @param new_val New value to be swapped
+ * @return 1 for success and 0 for fail
+ */
+#define ATOMIC_CAS_OP_128(atom, oper, old_val, new_val, val) \
+({ \
+	odp_u128_t _val; \
+	odp_atomic_u128_t *_atom = atom; \
+	odp_u128_t *_old_val = old_val; \
+	odp_u128_t _new_val = new_val; \
+	odp_u128_t *ptr = (odp_u128_t *)(_atom); \
+	register uint64_t old0 __asm__ ("x0"); \
+	register uint64_t old1 __asm__ ("x1"); \
+	register uint64_t new0 __asm__ ("x2"); \
+	register uint64_t new1 __asm__ ("x3"); \
+	old0 = (uint64_t)(_old_val)->u64[0]; \
+	old1 = (uint64_t)(_old_val)->u64[1]; \
+	new0 = (uint64_t)(_new_val).u64[0]; \
+	new1 = (uint64_t)(_new_val).u64[1]; \
+	__asm__ volatile(oper " %[old0], %[old1], %[new0], %[new1], [%[ptr]]" \
+			: [old0] "+r" (old0), [old1] "+r" (old1) \
+			: [new0] "r"  (new0), [new1] "r"  (new1), \
+			[ptr] "r" (ptr) \
+			: "memory"); \
+	_val.u64[0] = old0; \
+	_val.u64[1] = old1; \
+	val = _val; \
+})
+
+#define ATOMIC_CAS_OP_128_NO_ORDER(atom, old_value, new_value, val) \
+	ATOMIC_CAS_OP_128(atom, "casp", old_value, new_value, val)
+
+#define ATOMIC_CAS_OP_128_ACQ(atom, old_value, new_value, val) \
+	ATOMIC_CAS_OP_128(atom, "caspa", old_value, new_value, val)
+
+#define ATOMIC_CAS_OP_128_REL(atom, old_value, new_value, val) \
+	ATOMIC_CAS_OP_128(atom, "caspl", old_value, new_value, val)
+
+#define ATOMIC_CAS_OP_128_ACQ_REL(atom, old_value, new_value, val) \
+	ATOMIC_CAS_OP_128(atom, "caspal", old_value, new_value, val)
+
+#else
+
+/**
+ * @internal
+ * 128 bit store operation expression for the ATOMIC_OP macro
+ */
+#define ATOMIC_STORE_OP_128(new_val) \
+({ \
+	(_atom)->v = (new_val); \
+})
+
+/**
+ * @internal
+ * 128 bit CAS operation expression for the ATOMIC_OP macro
+ */
+#define ATOMIC_CAS_OP_128(ret_ptr, old_val, new_val) \
+({ \
+	int *_ret_ptr = ret_ptr; \
+	odp_u128_t *_old_val = old_val; \
+	odp_u128_t _new_val = new_val; \
+	if (((_atom)->v.u64[0] == (_old_val)->u64[0]) && \
+	    ((_atom)->v.u64[1] == (_old_val)->u64[1])) { \
+		(_atom)->v = (_new_val); \
+		*(_ret_ptr) = 1; \
+	} else { \
+		*(_ret_ptr) = 0; \
+	} \
+})
+
+/**
+ * @internal
+ * Helper macro for lock-based atomic operations on 128-bit integers
+ * @param[in,out] atom Pointer to the 128-bit atomic variable
+ * @param expr Expression used update the variable.
+ * @return The old value of the variable.
+ */
+#define ATOMIC_OP_128(atom, expr) \
+({ \
+	odp_u128_t _old_val; \
+	odp_atomic_u128_t *_atom = atom; \
+	/* Loop while lock is already taken, stop when lock becomes clear */ \
+	while (__atomic_test_and_set(&(_atom)->lock, __ATOMIC_ACQUIRE)) \
+		(void)0; \
+	_old_val = (_atom)->v; \
+	(expr); /* Perform whatever update is desired */ \
+	__atomic_clear(&(_atom)->lock, __ATOMIC_RELEASE); \
+	_old_val; /* Return old value */ \
+})
+
+#endif
```

Comment:
To reduce the amount of duplicate code these macros for lock-based operation could be moved to `platform/linux-generic/odp_atomic.c`.